internal/dataplane/kongstate/kongstate.go

```diff
@@ -690,3 +683,301 @@ func maybeLogKongIngressDeprecationError(logger logr.Logger, services []*corev1.
+			// TODO: we should generate multile entities if the plugin is attached to multiple services/routes/consumers.
+			if len(serviceIDs) > 0 {
+				parsedEntity[fieldName] = map[string]interface{}{
+					"id": serviceIDs[0],
+				}
+				logger.V(util.DebugLevel).Info("added ref to service", "service_id", serviceIDs[0])
+			}
+		case "routes":
+			routeIDs := lo.FilterMap(rels.Routes, func(r *Route, _ int) (string, bool) {
+				if err := r.FillID(workspace); err != nil {
+					return "", false
+				}
+				return *r.ID, true
+			})
+			if len(routeIDs) > 0 {
+				parsedEntity[fieldName] = map[string]interface{}{
+					"id": routeIDs[0],
+				}
+				logger.V(util.DebugLevel).Info("added ref to route", "route_id", routeIDs[0])
+			}
+		case "consumers":
+			consumerIDs := lo.FilterMap(rels.Consumers, func(c *Consumer, _ int) (string, bool) {
+				if err := c.FillID(workspace); err != nil {
+					return "", false
+				}
+				return *c.ID, true
+			})
+			if len(consumerIDs) > 0 {
+				parsedEntity[fieldName] = map[string]interface{}{
+					"id": consumerIDs[0],
+				}
+				logger.V(util.DebugLevel).Info("added ref to consumer", "consumer_id", consumerIDs[0])
+			}
+		}
+	}
+}
+
+// getServiceIDFromPluginRels returns the ID of the services which a plugin refers to in RelatedEntitiesRef.
+// It fills the IDs of services directly referred, and IDs of services where referred routes attaches to.
+func getServiceIDFromPluginRels(log logr.Logger, rels RelatedEntitiesRef, routeAttachedService map[string]*Service, workspace string) []string {
+	// Return IDs of directly referred services.
+	if len(rels.Services) > 0 {
+		return lo.FilterMap(rels.Services, func(s *Service, _ int) (string, bool) {
+			err := s.FillID(workspace)
+			if err != nil {
+				log.Error(err, "failed to fill ID for service")
+				return "", false
+			}
+			return *s.ID, true
+		},
+		)
+	}
+	// Returns IDs of services where the referred routes attaches.
+	if len(rels.Routes) > 0 {
+		serviceIDs := lo.FilterMap(
+			rels.Routes, func(r *Route, _ int) (string, bool) {
+				svc, ok := routeAttachedService[*r.Name]
+				if !ok {
+					return "", false
+				}
+				if err := svc.FillID(workspace); err != nil {
+					log.Error(err, "failed to fill ID for service")
+					return "", false
+				}
+				return *svc.ID, true
+			},
+		)
+		return lo.Uniq(serviceIDs)
+	}
+	return nil
+}
+
+// getPluginRelatedEntitiesRef gets services/routes/consumers referred by each plugin and returns the pointer to them.
+// It is for the custom entities to fill the IDs of the referred entities into their "foreign" fields.
+// It basically does the same thing as getPluginRelations but stores the pointers
+// because we need to call the FillID method of the entities to fetch the ID,
+// as Kong gateway requires IDs in the "foreign" fields (not other identifiers such as name.)
+func (ks *KongState) getPluginRelatedEntitiesRef(cacheStore store.Storer, log logr.Logger) PluginRelatedEntitiesRefs {
```

Comment:
This function look nearly identical to https://github.com/Kong/kubernetes-ingress-controller/blob/e7b3048dfa2b311009efb8dd3035cbec41f87b8a/internal/dataplane/kongstate/kongstate.go#L353-L422. Can we consider consolidating those 2 together? It can be done in a separate PR but less duplication would surely be a nice improvement here. WDYT?

Reply:
I wanted to consolidate the two places for getting plugin related entities, but this will make signifiant change to existing parts. I will create a new issue and finish that in a separate PR.

Reply:
Created #6115 to track it.